Para empezar a utilizar mongodb desde 0, vamos a partir realizando los siguientes pasos:


1.   Instalaremos $\texttt{mongodb}$
2.   Inicializaremos el proceso $\texttt{mongodb}$
3.   Instalaremos el comando $\texttt{gdown}$ para bajar la base de datos en formato json de google drive.
4.   Importaremos los datos por linea de comando. Dejaremos los datos en la base de datos $\texttt{local}$, en la colección $\texttt{pokedex}$, y le informaremos al comando que los datos vienen como una arreglo de jsons.




In [ ]:
!apt install mongodb
!service mongodb start
!pip install gdown
!gdown https://drive.google.com/uc?id=108aLE4mFE5bZYitFfX5c14Qj_aYA5MO_
!mongoimport --db local --collection pokedex --jsonArray < pokedex.json

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libpcap0.8 libstemmer0d libyaml-cpp0.5v5 mongo-tools mongodb-clients
  mongodb-server mongodb-server-core
The following NEW packages will be installed:
  libpcap0.8 libstemmer0d libyaml-cpp0.5v5 mongo-tools mongodb mongodb-clients
  mongodb-server mongodb-server-core
0 upgraded, 8 newly installed, 0 to remove and 37 not upgraded.
Need to get 53.1 MB of archives.
After this operation, 215 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libpcap0.8 amd64 1.8.1-6ubuntu1.18.04.2 [118 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 libstemmer0d amd64 0+svn585-1build1 [62.5 kB]
Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libyaml-cpp0.5v5 amd64 0.5.2-4ubuntu1 [150 kB]
Get:4 http://archive.ubuntu.com/ubuntu bionic/universe amd64 mongo-tools amd64 3.6.3-0ubunt

A continuación importaremos la librería de mongo para python, crearemos un cliente que se conectara a $\texttt{localhost}$ por defecto, nos conectaremos a la base local, y luego a la colección pokedex la cual almacenaremos en la variable $\texttt{collection}$.

In [ ]:
from pymongo import MongoClient
client = MongoClient()
client.list_database_names() # ['admin', 'local']
db = client['local']
collection = db['pokedex']

Para partir, podemos listar toda la colección usando el comando $\texttt{collection.find({})}$. Para no saturar la salida, nos limitaremos a mirar el primero de los registros.

In [ ]:
import pprint
for pokemon in collection.find({}).limit(1):
    pprint.pprint(pokemon)

{'Abilities': ['Overgrow', 'Chlorophyll'],
 'Attack': 49,
 'Capture Rate': 45,
 'Category': 'Seed Pokémon',
 'Defense': 49,
 'Egg Steps': 5120,
 'Exp Group': 'Medium Slow',
 'HP': 45,
 'Height (ft)': '2\'04"',
 'Height (m)': 0.7,
 'Id': '001',
 'Moves': {'Double-Edge': {'Accuracy': 100,
                           'Description': 'A reckless, life-risking tackle. '
                                          'This also damages the user quite a '
                                          'lot.',
                           'Effect %': '--',
                           'Level': 27,
                           'PP': 15,
                           'Power': 120,
                           'Type': 'Normal'},
           'Growl': {'Accuracy': 100,
                     'Description': 'The user growls in an endearing way, '
                                    'making opposing Pokémon less wary. This '
                                    'lowers their Attack stat.',
                     'Effect %': '--'

Fijense que aparte de los datos básicos del pokemon, también se presentan las habilidades (una lista de strings), y los movimientos (un arreglo asociativo/diccionario de los movimientos y sus detalles).

Si queremos ahora listar los pokemones que se llaman $\texttt{Charmander}$ podemos hacer la siguiente consulta: 

In [ ]:
for pokemon in collection.find({'Name': 'Charmander'}):
    pprint.pprint(pokemon)

{'Abilities': ['Blaze', 'Solar Power'],
 'Attack': 52,
 'Capture Rate': 45,
 'Category': 'Lizard Pokémon',
 'Defense': 43,
 'Egg Steps': 5120,
 'Exp Group': 'Medium Slow',
 'HP': 39,
 'Height (ft)': '2\'00"',
 'Height (m)': 0.6,
 'Id': '004',
 'Moves': {'Dragon Rage': {'Accuracy': 100,
                           'Description': 'This attack hits the target with a '
                                          'shock wave of pure rage. This '
                                          'attack always inflicts 40 HP '
                                          'damage.',
                           'Effect %': '--',
                           'Level': 16,
                           'PP': 10,
                           'Power': '??',
                           'Type': 'Dragon'},
           'Ember': {'Accuracy': 100,
                     'Description': 'The target is attacked with small flames. '
                                    'This may also leave the target with a '
                         

Y si queremos proyectar algunos de sus atributos lo hacemos tal como si fuera un diccionario:

In [ ]:
for pokemon in collection.find({'Name': 'Mew'}):
    print(pokemon['Speed'],pokemon['Abilities'])

100 ['Synchronize']


Recomendamos hechar un vistazo [aquí](https://docs.mongodb.com/manual/tutorial/query-documents/) para ver otros ejemplos de consultas que utilizarán en las siguientes preguntas:

#Pregunta 1
Encuentre los nombre de los pokemons que pesen mas de 150kg. 
[Ordene](https://www.w3schools.com/python/python_mongodb_sort.asp) los resultados por peso descendiente

#Pregunta 2
Encuentre los nombres de los pokemones ''Glass cannon'', que son los que tienen ataque sobre 120 y defensa bajo 50.

#Pregunta 3
Encuentre los pokemones "defensivos", que son los que tienen ya sea defensa, hp, o defensa especial (sp. defense) mayor o igual a 120.

#Pregunta 4
Seleccione los pokemones que tienen el movimiento $\texttt{Take Down}$ de nivel 15.

**Hint**: Puede consultar en diccionarios embebidos concatenando las llaves con un punto ".". 

Por ejemplo, para un dato:
```
[{
  a: {
    b: 'foo'
  }
}]
```
se puede consultar por los documentos tal que $\texttt{b}$ sea $\texttt{foo}$, escribiendo $\texttt{collection.find({'a.b': 'foo'})}$.

#Pregunta 5
Seleccione los pokemones que tienen el movimiento $\texttt{Take Down}$.

**Hint**: Puede testear si es que una llave existe utilizando [$exists](https://docs.mongodb.com/manual/reference/operator/query/exists/)

#Pregunta 6
1. Seleccione los pokemones que tienen la habilidad $\texttt{Telepathy}$


2. Seleccione los pokemones que tienen la habilidad $\texttt{Telepathy}$
 **y** $\texttt{Synchronize}$ (pueden tener mas habilidades aparte de estas dos. 
   
   **Hint**: puede utilizar el operador [$all](https://docs.mongodb.com/manual/tutorial/query-arrays/) . 

#Pregunta 7
Para la siguiente pregunta vamos a crear un índice sobre la columna $\texttt{Category}$.

In [ ]:
collection.create_index([('Category', 'text')])

'Category_text'

Usando el operador [$text](https://docs.mongodb.com/manual/text-search/) responda a las siguientes consultas.
1.  Devuelva los nombres y categorías de los pokemons tales que su categoría contenga ya sea la palabra **Sea** o **Gas**.

2. Devuelva los nombres y categorías de los pokemons tales que su categoría contenga la palabra **Sea** y no **Deep**.

3. Devuelva los nombres y categorías de los pokemons tales que su categoría contenga ambas palabras **Sea** y **Lion**.